1.0.0

bates64 opened the pull request:
Ready to merge as soon as https://github.com/decent-chat/decent/projects/7 is fully complete :)

Resolves #115
Fixes #234
Fixes #261
Fixes #276
Fixes #281
Fixes #305

bates64 committed:
Clarify mentions in API spec

bates64 committed:
Move mentions to its own endpoint

bates64 committed:
Use unix time for dates; add dots/periods to valid name spec

bates64 committed:
Remove 'authorization' and user.permissionLevel in favour of roles

bates64 committed:
Add permission list and color constants

bates64 committed:
Add roles

bates64 committed:
Style permission spoilers better

bates64 committed:
Convert errors into table and change some

Added: NO, FAILED, NOT_ALLOWED
Removed: MUST_BE_ADMIN

bates64 committed:
Restructure entire document

bates64 committed:
Unix time is not a year old

bates64 committed:
Use headers where appropriate

bates64 committed:
Add system messages

bates64 committed:
Add delete-user endpoint

bates64 committed:
Add MANAGE_USERS permission

bates64 committed:
Merge UPDATE_OTHERS and MANAGE_USERS

bates64 committed:
Fix typo

towerofnix committed:
Do permissions my way

towerofnix committed:
Tweak the rest of the docs to my permission system

towerofnix committed:
Put permission docs into their own section

towerofnix committed:
Move permissions object getting to /users/:id/permissions

towerofnix committed:
Clarify that emote shortcodes are ALWAYS unique

towerofnix committed:
Fix '{GET' typo

bates64 committed:
Remove messages.reactions; add messages.pinned

bates64 committed:
Re-add message/delete

towerofnix committed:
Add endpoints for updating channel-specific role permissions

towerofnix committed:
Change role.name to name, not string

towerofnix committed:
Revert "Change role.name to name, not string"

This reverts commit 29d661043b71fcdce13187258c371f0b29042071.

bates64 committed:
Add back emote/* events

bates64 committed:
Add serverSettings.iconURL & server setting update event

Fixes #278.

towerofnix committed:
Make emote shortcodes strings

towerofnix committed:
Be more specific about special role IDs

Commit:
Merge pull request #275 from decent-chat/permissions-spec-2

Redo 1.0.0 permission spec

bates64 committed:
POST -> PATCH /api/settings

bates64 committed:
Add sendSystemMessages permission

bates64 committed:
Delete packages/client/src/

Rest in pizza.

bates64 committed:
Merge branch 'master' into preact

bates64 committed:
Setup Preact build steps, add PostCSS autoprefixer

bates64 committed:
Add FF to browserslist

bates64 committed:
Support all browser versions released in the last year

bates64 committed:
Add server dropdown

PullJosh committed:
Add basic channel list and modals

PullJosh committed:
Track UI state per-server; use new per-server UI state to track active channel

bates64 committed:
Fix codestyle

@PullJosh - use single quotes!

bates64 committed:
Use portals for modals

PullJosh committed:
Add semi-functional UserList. Still need to make it auto-update!

bates64 committed:
Refactor modals

bates64 committed:
Merge branch 'preact-user-list' into preact

bates64 committed:
Fix crash oops

bates64 committed:
Close modals when clicking behind them

bates64 committed:
Add <Icon>

bates64 committed:
Add server pool

PullJosh committed:
Add current user status; need client + pool event for login/out

bates64 committed:
Add decent.js Client login/logout events

bates64 committed:
Add decent.js client.register()

bates64 committed:
Add login/logout to client

bates64 committed:
Use a <Modal.Async/> for join server modal

bates64 committed:
Use --no-bg 'Cancel' buttons

bates64 committed:
Remove 'logged in as <username>' extra spacing

PullJosh committed:
Horizontal bar loading for modals

PullJosh committed:
Merge branch 'preact' of github.com:decent-chat/decent into preact

bates64 committed:
Add disconnected toast

bates64 committed:
Merge remote-tracking branch 'origin/preact' into preact

bates64 committed:
Use Preact's debug mode in development

bates64 committed:
Add initial message scroller

PullJosh committed:
Fix timeago

bates64 committed:
Listen for messages

bates64 committed:
Merge remote-tracking branch 'origin/preact' into preact

bates64 committed:
Add scrollback/forward

bates64 committed:
Add message action handlers

bates64 committed:
Fix new message receiving

PullJosh committed:
Allow sending messages (add MessageEditor)

bates64 committed:
Bump message scroller max length to 50 messages

bates64 committed:
Fix seeing double

bates64 committed:
Clean up message editor

PullJosh committed:
First attempt at image uploads. Paging @heyitsmeuralex to investigate.

PullJosh committed:
Merge in work from alex

bates64 committed:
Maybe fix scrolledToBottom detection a bit

bates64 committed:
Merge remote-tracking branch 'origin/preact' into preact

bates64 committed:
Remove rogue var self = this

bates64 committed:
Fix client.uploadImage

bates64 committed:
Display progress bar for image uploading

bates64 committed:
Clarify sessions and permissions, add REPEATED_PARAMETERS

bates64 committed:
Move server properties to /api

PullJosh committed:
Make addition of messages at top/bottom of list be completely seamless

bates64 committed:
Make textarea better

bates64 committed:
Merge remote-tracking branch 'origin/preact' into preact

PullJosh committed:
Make modal close button visible

bates64 committed:
Fix message editor textarea autosizing bugs

bates64 committed:
Fix <Modal/> bugs

bates64 committed:
Implement message image upload button

PullJosh committed:
Reorganize components to be infinitely nestable

bates64 committed:
Greatly simplify require(..) calls using a custom Babel plugin

Commit:
Merge pull request #292 from decent-chat/organize-components

Organize Components

bates64 committed:
Fix <InfiniteScroll/>

bates64 committed:
Add 'npm run serve'

bates64 committed:
Fix message editor sizing once again

bates64 committed:
Listen for message deletion

PullJosh committed:
Always show top bar of Messages component, even while loading.

towerofnix committed:
Implement unix timestamps

towerofnix committed:
Add type property to messages

bates64 committed:
Merge remote-tracking branch 'origin/impl-1.0.0-msg-types' into 1.0.0

bates64 committed:
Merge remote-tracking branch 'origin/impl-1.0.0-dates' into 1.0.0

towerofnix committed:
Merge branch 'master' into 1.0.0

towerofnix committed:
Change iconURL -> iconPath, POST settings -> PATCH, add WS event

towerofnix committed:
Remove /api/properties, add useSecureProtocol, tests

towerofnix committed:
Make client not fetch /api/properties

This disables authorized users, but that isn't a big deal, since those
are to be changed to use permissions/etc later.

towerofnix committed:
Change iconPath back to iconURL

Commit:
Merge pull request #298 from decent-chat/impl-1.0.0-settings

Implement 1.0.0-spec settings

bates64 committed:
Use dateEdited and dateCreated for messages

bates64 committed:
Store millisecond-accurate dates

Fixes #297. The API currently Math.floor()s the dates before outputting them, however the spec should specify whether they should *always* be integers or *can* be floats (eg. 1500000.25).

bates64 committed:
Don't floor datetime values

towerofnix committed:
Merge pull request #299 from decent-chat/impl-1.0.0-better-dates

1.0.0 impl: various datetime fixes

towerofnix committed:
Remove auth stuff; add POST /api/roles

Note that POST /api/roles is *definitely* not done yet.

towerofnix committed:
Add GET /api/roles

towerofnix committed:
More role stuff

towerofnix committed:
Add more tests; require role names to be at least one char

towerofnix committed:
Add DELETE /api/roles/:id

towerofnix committed:
Be more strict about what can be done to internal and guest roles

towerofnix committed:
Be more strict in validate.permissionsObject; add test for it

towerofnix committed:
Very very very begin user permissions

towerofnix committed:
Fun documentation edits

towerofnix committed:
Tweaks and things

towerofnix committed:
Get rid of lots and lots of authorized stuff; add lots and lots of roles stuff

towerofnix committed:
Remove AUTHORIZATION_ERROR & mention of auth in docs

towerofnix committed:
Add doc notes regarding role pri. order

towerofnix committed:
Get rid of _owner; clean up other things

towerofnix committed:
Implement Owner role in a much saner way

towerofnix committed:
Update make-admin CLI command

towerofnix committed:
Begin work on permissions actually doing stuff

I think this commit breaks some be-admin-related tests, but that's not
really relevant.

towerofnix committed:
Add a test, fix another test

towerofnix committed:
In makeChannel test util, don't reference old makeAdmin function

towerofnix committed:
Change everything, make channel API use permissions

towerofnix committed:
Clean up channel test code

towerofnix committed:
Make emote API use permissions

towerofnix committed:
Generally update PATCH /api/settings

towerofnix committed:
Deupdate PATCH /api/settings

towerofnix committed:
Restore NOT_ALLOWED, sorry!!!

towerofnix committed:
Fix typo in api.md

towerofnix committed:
Typecheck all the things

towerofnix committed:
Use cleaner typechecking in settings area

TODO: this same type of code in the user typechecking section!

towerofnix committed:
Update messages API to use permissions

bates64 committed:
Update to work with 1.0.0 servers

bates64 committed:
Fix package.json except @decent/server's

bates64 committed:
Delete packages/client/src/

bates64 committed:
Merge 'preact' into '1.0.0'

bates64 committed:
Bump all packages to v1.0.0

bates64 committed:
Merge remote-tracking branch 'origin/impl-1.0.0-permissions-roles' into 1.0.0

bates64 committed:
Add null states + various fixes

* Fixed message inputs autofilling weirdly
* Added null state for when you're not in a channel
* Added null state for when you're have no hosts in the hosts pool (display connect-server modal)
* <Modal>s can now specify cancellable=false

bates64 committed:
Fix some things, rename 'Owner' role to 'Admin' and change semantics

bates64 committed:
Add dark theme back, some design tweaks

bates64 committed:
Use Gravatar (fix #305) and fix all broken tests

bates64 committed:
Use a 16-color scheme everywhere

bates64 committed:
Major design changes

bates64 committed:
Fix create channel button

bates64 committed:
Fix decent.js Channel#getMessages typecheck issue

bates64 committed:
Fix <Toast> styling

bates64 committed:
Use glob postcss imports

bates64 committed:
Move CSS files around, small tweaks, add modal/dropdown animations

bates64 committed:
Add inline message editing (resolves #115)

bates64 committed:
Fix issue where message actions didn't appear after login

bates64 committed:
Fix some bugs

bates64 committed:
Eagerly load more messages before the actual top/bottom is reached

bates64 committed:
Add unread badges + document.title

towerofnix committed:
Disable reactions endpoint

towerofnix committed:
Kill api.md in favor of decent-chat/spec

towerofnix committed:
Add util functions for future role/permission stuff + tests

towerofnix committed:
Remove _guest role

bates64 committed:
decent.js: add Roles

bates64 committed:
Merge remote-tracking branch 'origin/1.0.0' into 1.0.0

bates64 committed:
Add user popups

Also updated <Dropdown> component so they never appear offscreen.

bates64 committed:
Persist server list between page refreshes

bates64 committed:
Fix potential activeServer bug

towerofnix committed:
Add GET /api/users/:id/roles

towerofnix committed:
WIP POST user roles -- middleware stuff, improve error messages

towerofnix committed:
WIP POST user roles -- start fleshing endpoint

towerofnix committed:
Finish POSTing user roles

towerofnix committed:
Add DELETE /users/:userID/roles/:roleID

bates64 committed:
Add disconnected server UI

Also fixed a few bugs, and made the state of the server quick switcher (ServerList) persist over reloads.

bates64 committed:
Merge remote-tracking branch 'origin/1.0.0' into 1.0.0

towerofnix commented on packages/client/js/storage.js:
```diff
@@ -0,0 +1,18 @@
+  const str = storage.getItem(key)
+  const value = str ? JSON.parse(str) : defaultValue
+
+  console.log(`Storage: loaded ${key} =`, value)
```
Don't keep these in the final 1.0.0!

bates64 committed:
Remove logs

Before I forget

bates64 committed:
Fix client - never mistakenly set the active server to -1

bates64 committed:
Fix broken 'connect to server' modal

bates64 committed:
Add build:favicon

bates64 committed:
Delete messages when edited to have no content

Fixes #311.

towerofnix committed:
WIP - channel permissions

Just a couple mostly-dummy endpoints for now (and no tests). Still want
to do some needed work for permissions on the other endpoints first.

towerofnix committed:
Merge branch '1.0.0' of https://github.com/decent-chat/decent into 1.0.0

heyitsmeuralex commented on packages/server/api.js:
```diff
@@ -742,6 +742,30 @@ module.exports = async function attachAPI(app, {wss, db, dbDir}) {
+      const { channel } = request[middleware.vars]
+    }
+  ])
+  */
```
um

towerofnix replied:
WIP, of course.

towerofnix committed:
Add getHighestRoleOfUser util function

towerofnix committed:
Update the way new roles are prioritized per spec

towerofnix committed:
Don't allow reordering roles above own highest role

towerofnix committed:
Test for manageRoles permission check

towerofnix committed:
Don't let roles be reordered such that you lose manageRoles

towerofnix committed:
Make POST /api/roles match spec

towerofnix committed:
Woops, also check for manageRoles

bates64 committed:
Add support for system messages to server & decent.js

bates64 committed:
Remove authorFlair from messages to match spec

towerofnix commented on packages/server/test/api-messages.js:
```diff
@@ -47,6 +47,21 @@ test('POST /api/messages', t => {
+    } catch (error) {
+      t.is(error.code, 'NOT_ALLOWED')
+      t.is(error.missingPermission, 'sendSystemMessages')
+    }
```
@heyitsmeuralex You still need to update `serialize.message`, and you should update `test/serialize.js` too.

bates64 committed:
Serialize system messages properly

towerofnix reviewed (commented):
Just a nit and a question re: flair test. System messages look good otherwise!

heyitsmeuralex commented on packages/server/serialize.js:
```diff
@@ -7,9 +7,9 @@ module.exports = function makeSerializers({util, db}) {
-      authorAvatarURL: emailToAvatarURL(m.authorEmail || m.authorID),
+      authorUsername: m.type === 'system' ? null : m.authorUsername,
+      authorID: m.type === 'system' ? null : m.authorID,
+      authorAvatarURL: m.type === 'system' ? null : emailToAvatarURL(m.authorEmail || m.authorID),
```
If we ever add more types they'll likely be tied to users too - `system` is the special case, not `user`.
